File: backend/api/place_groups.py
"""The eight place groups (מחנה/מוצב/צומת/מחלף/שטח אש/כביש/מחסום/אחר) and their display prefixes.

A place's full display string (e.g. "מ. אלפורן") is always what the API sends
and receives — the wire contract never changed. Internally, ``api.db.Database``
stores each place as a numeric id in a registry (``places.json``:
``{id: {"name": base_name, "group": group_key}}``) so that two distinct places
which happen to share a base name after stripping their prefix (e.g. "מ. גולני"
the camp and "מחלף גולני" the interchange) can never collide — this module is
what ``Database`` uses to parse a display string's prefix into
``(group, base_name)`` when resolving/minting that id, and to reconstruct the
display string back from a registry entry.

``frontend/src/utils/placeGroups.js`` is a separate, independent copy of this
same group/prefix vocabulary, used purely for *client-side* classification
(HomePage's type-filters, and deciding when the route editor's "which group?"
popover should appear) — since the frontend only ever sees full display
strings, it never needs to resolve or store an id at all. Keeping the two
copies conceptually aligned matters for UX consistency (so the client doesn't
prompt for a group the server would have parsed automatically, or vice versa),
but any drift is low-stakes: the server remains the sole authority for what
actually gets persisted.
"""

import re

DEFAULT_GROUP = "other"

# Key order matters: it is the group filter-chip render order on the frontend,
# kept identical there.
GROUPS = {
    "camp": {"label": "מחנה", "prefix": "מ."},
    "post": {"label": "מוצב", "prefix": "מוצב"},
    "junction": {"label": "צומת", "prefix": "צ."},
    "interchange": {"label": "מחלף", "prefix": "מחלף"},
    "firing_zone": {"label": "שטח אש", "prefix": "ש.א"},
    "road": {"label": "כביש", "prefix": "כביש"},
    "checkpoint": {"label": "מחסום", "prefix": "מחסום"},
    DEFAULT_GROUP: {"label": "אחר", "prefix": None},
}

GROUP_KEYS = set(GROUPS)

# One regex per group with a real prefix (skips "other", which has none). Order
# matters: "מוצב" must be tried before a bare "מ." pattern could ever partially
# match it, though the patterns below are specific enough that order is not
# actually load-bearing today — kept explicit regardless for parity with the
# frontend list's order.
PREFIX_PATTERNS = [
    ("camp", re.compile(r"^מ(?:\.\s*|\s+)(?=\S)")),
    ("post", re.compile(r"^מוצב\s+(?=\S)")),
    ("junction", re.compile(r"^צ(?:\.\s*|\s+)(?=\S)")),
    ("interchange", re.compile(r"^מחלף\s*(?=\S)")),
    ("firing_zone", re.compile(r"^ש[.\s]?א\s*(?=\S)")),
    ("road", re.compile(r"^כביש\s+(?=\S)")),
    ("checkpoint", re.compile(r"^מחסום\s+(?=\S)")),
]


def parse_prefixed_name(text):
    """Split a typed/stored string into ``(group_key, base_name)`` if it starts
    with one of the five recognized prefixes, else ``None``.

    The caller decides what "no match" means: the one-time migration defaults it
    to :data:`DEFAULT_GROUP`, while the live route editor asks the user instead
    of guessing.
    """
    trimmed = (text or "").strip()
    for group_key, pattern in PREFIX_PATTERNS:
        match = pattern.match(trimmed)
        if match:
            base = trimmed[match.end():].strip()
            if base:
                return group_key, base
    return None


def format_place(base_name, group_key):
    """The prefixed display string for a base name + group.

    Used by the migration command's log output; the frontend module
    (``placeGroups.js``) is the real display authority for the UI.
    """
    prefix = GROUPS.get(group_key, GROUPS[DEFAULT_GROUP])["prefix"]
    return f"{prefix} {base_name}" if prefix else base_name

File: frontend/src/utils/placeGroups.js
/**
 * The eight place groups (מחנה/מוצב/צומת/מחלף/שטח אש/כביש/מחסום/אחר), each with
 * a display prefix. The backend always sends/receives the full display string (e.g.
 * "צ. גומא") — a place's group is never a separate field on the wire, so this
 * module classifies purely by reading a name's own prefix back, the same way
 * everywhere in the app that needs a place's group does it.
 *
 * `test` here should stay conceptually aligned with `backend/api/place_groups.py`'s
 * `PREFIX_PATTERNS` — both parse the same real-world strings — though drift
 * here is lower-stakes than it would be for stored data: the worst case is a
 * misclassification for filtering/search, or an unnecessary/missing
 * "assign a group" prompt when adding a stop, never data corruption (the
 * backend is the sole authority for what actually gets persisted).
 */

export const DEFAULT_GROUP = "other";

export const PLACE_GROUPS = [
  { key: "camp", label: "מחנה", prefix: "מ.", test: /^מ(?:\.\s*|\s+)(?=\S)/ },
  { key: "post", label: "מוצב", prefix: "מוצב", test: /^מוצב\s+(?=\S)/ },
  { key: "junction", label: "צומת", prefix: "צ.", test: /^צ(?:\.\s*|\s+)(?=\S)/ },
  { key: "interchange", label: "מחלף", prefix: "מחלף", test: /^מחלף\s*(?=\S)/ },
  { key: "firing_zone", label: "שטח אש", prefix: "ש.א", test: /^ש[.\s]?א\s*(?=\S)/ },
  { key: "road", label: "כביש", prefix: "כביש", test: /^כביש\s+(?=\S)/ },
  { key: "checkpoint", label: "מחסום", prefix: "מחסום", test: /^מחסום\s+(?=\S)/ },
  { key: DEFAULT_GROUP, label: "אחר", prefix: null, test: null },
];

const BY_KEY = new Map(PLACE_GROUPS.map((g) => [g.key, g]));

/** The group a display/typed name's own prefix implies — always a real group
 * key, "אחר" (never null) when nothing matches. */
export function classifyPlace(name) {
  const trimmed = (name || "").trim();
  const group = PLACE_GROUPS.find((g) => g.test && g.test.test(trimmed));
  return group ? group.key : DEFAULT_GROUP;
}

/** Strip a *recognized* (non-"אחר") prefix off typed text, returning
 * `{ group, baseName }`, or `null` if nothing matches — the caller (the route
 * editor's new-stop flow) must ask the user for a group in that case, rather
 * than silently defaulting to "אחר". */
export function parseTypedPlace(text) {
  const trimmed = (text || "").trim();
  for (const g of PLACE_GROUPS) {
    if (g.test && g.test.test(trimmed)) {
      return { group: g.key, baseName: trimmed.replace(g.test, "").trim() };
    }
  }
  return null;
}

/** The full display string for a base name + group — used to *construct* a
 * stop's committed text after the user picks a group in the ask-popover, the
 * same as if they'd typed the prefix themselves. */
export function formatPlace(baseName, groupKey) {
  const group = BY_KEY.get(groupKey);
  return group?.prefix ? `${group.prefix} ${baseName}` : baseName;
}

export function groupLabel(key) {
  return (BY_KEY.get(key) ?? BY_KEY.get(DEFAULT_GROUP)).label;
}

/** Fixed categorical palette for BrainPage's "color by group" node-fill mode
 * (see GraphView.jsx) — kept here so the graph and its legend never drift. */
export const GROUP_COLORS = {
  camp: "#fb923c",
  post: "#c084fc",
  junction: "#4ade80",
  interchange: "#facc15",
  firing_zone: "#f87171",
  road: "#60a5fa",
  checkpoint: "#e879f9",
  other: "#94a3b8",
};
